// @vercel/cli-exec/dist/index.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var src_exports = {};
__export(src_exports, {
  VercelCliError: () => VercelCliError,
  clearVercelCliCache: () => clearVercelCliCache,
  execVercelCli: () => execVercelCli,
  findVercelCli: () => findVercelCli
});
module.exports = __toCommonJS(src_exports);
var import_node_fs = require("node:fs");
var import_node_path = __toESM(require("node:path"));
var import_execa = __toESM(require("execa"));
class VercelCliError extends Error {
  constructor(options) {
    super(options.message);
    this.name = "VercelCliError";
    this.code = options.code;
    this.invocation = options.invocation;
    this.stdout = options.stdout;
    this.stderr = options.stderr;
    this.exitCode = options.exitCode;
    if (options.cause !== void 0) {
      this.cause = options.cause;
    }
  }
}
function findVercelCli(options = {}) {
  const cwd = import_node_path.default.resolve(options.cwd ?? process.cwd());
  const pathValue = options.path ?? getEnvPath(process.env);
  const cacheKey = getCliInvocationCacheKey(cwd, pathValue);
  if (cliInvocationCache.has(cacheKey)) {
    return cliInvocationCache.get(cacheKey) ?? null;
  }
  const invocation = resolveCliInvocation(cwd, pathValue);
  cliInvocationCache.set(cacheKey, invocation);
  return invocation;
}
function clearVercelCliCache() {
  cliInvocationCache.clear();
}
async function execVercelCli(args, options = {}) {
  const cwd = import_node_path.default.resolve(options.cwd ?? process.cwd());
  assertValidCwd(cwd);
  const env = mergeExecEnv(options.env);
  const pathValue = getEnvPath(env);
  const cacheKey = getCliInvocationCacheKey(cwd, pathValue);
  for (let attempt = 0; attempt < 2; attempt++) {
    const invocation = findVercelCli({ cwd, path: pathValue });
    if (!invocation) {
      throw new VercelCliError({
        code: "VERCEL_CLI_NOT_FOUND",
        message: "Unable to find a usable Vercel CLI installation."
      });
    }
    try {
      const execaOptions = {
        input: options.input,
        stdio: options.stdio,
        stdin: options.stdin,
        stdout: options.stdout,
        stderr: options.stderr,
        timeout: options.timeout,
        cwd,
        env: prependLocalBinsToEnvPath(cwd, env),
        windowsHide: true
      };
      if (options.signal) {
        execaOptions.signal = options.signal;
      }
      const { stdout, stderr } = await (0, import_execa.default)(
        invocation.command,
        [...invocation.commandArgs, ...args],
        execaOptions
      );
      return { stdout, stderr, invocation };
    } catch (error) {
      const wrappedError = toVercelCliError(invocation, error);
      if (wrappedError.code === "VERCEL_CLI_NOT_FOUND" && attempt === 0) {
        cliInvocationCache.delete(cacheKey);
        continue;
      }
      throw wrappedError;
    }
  }
  throw new VercelCliError({
    code: "VERCEL_CLI_NOT_FOUND",
    message: "Unable to find a usable Vercel CLI installation."
  });
}
const cliInvocationCache = /* @__PURE__ */ new Map();
function getVercelCommandNames() {
  const commandBases = ["vercel"];
  if (process.platform !== "win32") {
    return commandBases;
  }
  const extensions = [".cmd", ".exe", ""];
  return commandBases.flatMap(
    (command) => extensions.map((extension) => `${command}${extension}`)
  );
}
function getAncestorDirectories(cwd) {
  const directories = [];
  let current = import_node_path.default.resolve(cwd);
  while (true) {
    directories.push(current);
    const parent = import_node_path.default.dirname(current);
    if (parent === current) {
      return directories;
    }
    current = parent;
  }
}
function getCanonicalPath(filePath) {
  try {
    return (0, import_node_fs.realpathSync)(filePath);
  } catch {
    return filePath;
  }
}
function getLocalBinDirectories(cwd) {
  const searchRoot = getCanonicalPath(import_node_path.default.resolve(cwd));
  return getAncestorDirectories(searchRoot).map(
    (directory) => import_node_path.default.join(directory, "node_modules", ".bin")
  );
}
function isNodeScript(filePath) {
  return [".js", ".cjs", ".mjs"].includes(import_node_path.default.extname(filePath));
}
function splitPath(pathValue) {
  return pathValue.split(import_node_path.default.delimiter).filter(Boolean);
}
function getEnvPath(env = process.env) {
  if (process.platform !== "win32") {
    return env.PATH ?? "";
  }
  const pathKeys = Object.keys(env).filter((key) => key.toLowerCase() === "path");
  for (let index = pathKeys.length - 1; index >= 0; index--) {
    const value = env[pathKeys[index]];
    if (value !== void 0) {
      return value;
    }
  }
  return "";
}
function setEnvPath(env = process.env, pathValue) {
  if (process.platform !== "win32") {
    return {
      ...env,
      PATH: pathValue
    };
  }
  const normalizedEnv = { ...env };
  for (const key of Object.keys(normalizedEnv)) {
    if (key !== "PATH" && key.toLowerCase() === "path") {
      delete normalizedEnv[key];
    }
  }
  normalizedEnv.PATH = pathValue;
  return normalizedEnv;
}
function mergeExecEnv(env) {
  if (!env) {
    return process.env;
  }
  return { ...process.env, ...env };
}
function findCommandInPath(command, pathValue, cwd) {
  for (const directory of splitPath(pathValue)) {
    const candidateDirectory = import_node_path.default.isAbsolute(directory) ? directory : import_node_path.default.resolve(cwd, directory);
    const candidate = import_node_path.default.join(candidateDirectory, command);
    try {
      (0, import_node_fs.accessSync)(
        candidate,
        process.platform === "win32" ? import_node_fs.constants.F_OK : import_node_fs.constants.F_OK | import_node_fs.constants.X_OK
      );
      if (!(0, import_node_fs.statSync)(candidate).isFile()) {
        continue;
      }
      return { path: candidate, realPath: (0, import_node_fs.realpathSync)(candidate) };
    } catch {
    }
  }
  return null;
}
function getCliInvocationCacheKey(cwd, pathValue) {
  return `${cwd}\0${pathValue}`;
}
function isLocalBinPath(cwd, filePath) {
  const resolvedFilePath = import_node_path.default.resolve(filePath);
  let canonicalFilePath = resolvedFilePath;
  try {
    canonicalFilePath = import_node_path.default.join(
      (0, import_node_fs.realpathSync)(import_node_path.default.dirname(resolvedFilePath)),
      import_node_path.default.basename(resolvedFilePath)
    );
  } catch {
  }
  return getLocalBinDirectories(cwd).some((localBinDirectory) => {
    try {
      localBinDirectory = (0, import_node_fs.realpathSync)(localBinDirectory);
    } catch {
    }
    return canonicalFilePath.startsWith(`${localBinDirectory}${import_node_path.default.sep}`);
  });
}
function prependLocalBinsToPath(cwd, pathValue = "") {
  return prependPathEntries(pathValue, getLocalBinDirectories(cwd));
}
function prependPathEntries(pathValue, directories) {
  const pathParts = pathValue.split(import_node_path.default.delimiter).filter(Boolean);
  const prepended = [];
  for (const directory of directories) {
    if (!pathParts.includes(directory) && !prepended.includes(directory)) {
      prepended.push(directory);
    }
  }
  if (prepended.length === 0) {
    return pathValue;
  }
  return pathValue === "" || pathValue === import_node_path.default.delimiter ? `${prepended.join(import_node_path.default.delimiter)}${pathValue}` : [...prepended, pathValue].join(import_node_path.default.delimiter);
}
function prependLocalBinsToEnvPath(cwd, env = process.env) {
  const localPath = prependLocalBinsToPath(cwd, getEnvPath(env));
  return setEnvPath(
    env,
    prependPathEntries(localPath, [import_node_path.default.dirname(process.execPath)])
  );
}
function resolveCliInvocation(cwd, pathValue) {
  const resolvedPath = prependLocalBinsToPath(cwd, pathValue);
  for (const command of getVercelCommandNames()) {
    const resolvedCommand = findCommandInPath(command, resolvedPath, cwd);
    if (!resolvedCommand) {
      continue;
    }
    const source = isLocalBinPath(cwd, resolvedCommand.path) ? "local-bin" : "path";
    if (isNodeScript(resolvedCommand.realPath)) {
      return {
        command: process.execPath,
        commandArgs: [resolvedCommand.realPath],
        source
      };
    }
    return {
      command: resolvedCommand.realPath,
      commandArgs: [],
      source
    };
  }
  return null;
}
function assertValidCwd(cwd) {
  try {
    if (!(0, import_node_fs.statSync)(cwd).isDirectory()) {
      throw new Error("not a directory");
    }
  } catch {
    throw new VercelCliError({
      code: "VERCEL_CLI_INVALID_CWD",
      message: `Working directory ${JSON.stringify(cwd)} does not exist or is not a directory.`
    });
  }
}
function toVercelCliError(invocation, error) {
  if (typeof error === "object" && error !== null) {
    const execaError = error;
    if (execaError.code === "ENOENT") {
      return new VercelCliError({
        code: "VERCEL_CLI_NOT_FOUND",
        message: `Unable to find Vercel CLI command ${JSON.stringify(invocation.command)}.`,
        invocation,
        cause: error
      });
    }
    if (execaError.code === "EACCES" || execaError.code === "EPERM") {
      return new VercelCliError({
        code: "VERCEL_CLI_PERMISSION_DENIED",
        message: `Permission denied while executing Vercel CLI command ${JSON.stringify(invocation.command)}.`,
        invocation,
        cause: error
      });
    }
    if (execaError.timedOut) {
      return new VercelCliError({
        code: "VERCEL_CLI_TIMED_OUT",
        message: `Timed out while executing Vercel CLI command ${JSON.stringify(invocation.command)}.`,
        invocation,
        stdout: execaError.stdout,
        stderr: execaError.stderr,
        cause: error
      });
    }
    if (execaError.isCanceled) {
      return new VercelCliError({
        code: "VERCEL_CLI_CANCELED",
        message: `Canceled while executing Vercel CLI command ${JSON.stringify(invocation.command)}.`,
        invocation,
        stdout: execaError.stdout,
        stderr: execaError.stderr,
        cause: error
      });
    }
    if (execaError.signal) {
      return new VercelCliError({
        code: "VERCEL_CLI_SIGNALED",
        message: `Vercel CLI command ${JSON.stringify(invocation.command)} exited due to signal ${execaError.signal}.`,
        invocation,
        stdout: execaError.stdout,
        stderr: execaError.stderr,
        cause: error
      });
    }
    if (typeof execaError.exitCode === "number") {
      return new VercelCliError({
        code: "VERCEL_CLI_ERRORED",
        message: execaError.shortMessage ?? execaError.message ?? `Vercel CLI command ${JSON.stringify(invocation.command)} exited with code ${execaError.exitCode}.`,
        invocation,
        stdout: execaError.stdout,
        stderr: execaError.stderr,
        exitCode: execaError.exitCode,
        cause: error
      });
    }
  }
  return new VercelCliError({
    code: "VERCEL_CLI_EXEC_FAILED",
    message: `Could not execute Vercel CLI command ${JSON.stringify(invocation.command)}.`,
    invocation,
    cause: error
  });
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  VercelCliError,
  clearVercelCliCache,
  execVercelCli,
  findVercelCli
});
